### ISSUE TITLE
Add Parameter to ConvertTo-Json to ignore unsupported properties

### ISSUE THREAD

**Issue by mike-the-automator** [PowerShell/PowerShell#5749]
I'd like to propose an additional parameter for the ConvertTo-Json cmdlet that specifies the caller would like the cmdlet to ignore properties of the InputObject that cannot be converted to JSON successfully.  Currently, the behavior of the cmdlet is to throw an exception with a message similar to the following

`"ConvertTo-Json : The type 'System.Collections.ListDictionaryInternal' is not supported for
serialization or "deserialization of a dictionary. Keys must be strings.`

My proposal is that if the `-SkipIncompatibleProperties` (parm name feedback appreciated :-) ) parameter is specified, then the cmdlet would silently skip over these properties and not throw any exceptions.

I've been frustrated by this error several times in my scripting in situations where I don't really care about the particular property that triggered the error; I would just like a quick and easy serialized representation of the object.  In the most recent instance, I needed a way to record exception info in a VARCHAR column in SQL Server.  The property of the System.Exception that triggered this error was the "Data" field which I didn't really care about, but it prevented me from easily using the cmdlet to serialize the exception object.

As an aside, it looks like there is a proposal by @KirkMunro to [move the JSON cmdlets to their own module.](https://github.com/PowerShell/PowerShell/issues/3995)  I understand it may make sense to wait on this feature until the module split occurs.

**Comment by bergmeister** [comment]
I recently added a new -`AsHashtable` switch to it to cover cases where the Json could not be parsed in the past due to the limitations if `PsObject`. Can you please give an example where the new switch would throw as well?

**Comment by bergmeister** [comment]
@mike-the-automator Does 
 `-ErrorAction SilentlyContinue` solve your specific case(s)? If not then it might be the better solution to rather tweak the error behaviour instead of adding a new parameter.

**Comment by mike-the-automator** [comment]
Thanks for your response.  I tested `-ErrorAction SilentlyContinue` and that did not seem to solve my problem.  

Transcript:
```
PowerShell v6.0.0-rc.2
Copyright (c) Microsoft Corporation. All rights reserved.

https://aka.ms/pscore6-docs
Type 'help' to get help.

PS C:\Program Files\PowerShell6RC2\6.0.0-rc.2> 1/0
Attempted to divide by zero.
At line:1 char:1
+ 1/0
+ ~~~
+ CategoryInfo          : NotSpecified: (:) [], RuntimeException
+ FullyQualifiedErrorId : RuntimeException

PS C:\Program Files\PowerShell6RC2\6.0.0-rc.2> $j = ConvertTo-Json $Error[0] -ErrorAction SilentlyContinue
ConvertTo-Json : The type 'System.Collections.ListDictionaryInternal' is not supported for serialization or deserialization of a dictionary. Keys must be strings.
At line:1 char:6
+ $j = ConvertTo-Json $Error[0] -ErrorAction SilentlyContinue
+      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (System.Collecti...tionaryInternal:ListDictionaryInternal) [ConvertTo-Json], InvalidOperationException
+ FullyQualifiedErrorId : NonStringKeyInDictionary,Microsoft.PowerShell.Commands.ConvertToJsonCommand

PS C:\Program Files\PowerShell6RC2\6.0.0-rc.2> $j -eq $null
True

````
 In the scenario I envision, when the new parameter is supplied there would be no error thrown and the value of the variable `$j` would be a JSON representation of the exception objects, ignoring any fields that would have triggered the InvalidOperation exception.

**Comment by mike-the-automator** [comment]
@bergmeister I have mulled over the suggestion to just change the behavior for `-ErrorAction SilentlyContinue` and my impression is that would be a counter intuitive way to achieve my goal.  It would be a breaking change for users because a scenario that used to generate an Exception would no longer do so.  

Also, silently discarding data without being explicitly told to do so feels _wrong_.  It almost feel more like an occasion to use `-Force` as more of a "No, really this **is** what I want to do" kind of parameter.

Another means to get the result I want would be to just create a cmdlet to strip non-JSON compliant fields of an object that would result in this error and put it in front of ConvertTo-Json in the pipeline, e.g.

```

$j = $Error[0] | Remove-JsonIncompatibleProperties | ConvertTo-Json

```

I could always create a convenience wrapper-function to eliminate the need for a long pipeline (I suck at naming things).  Given that perspective I can understand if the consensus is that the complexity/utility ratio doesn't pan out for this feature request, but I'd like to know what others think.

**Comment by FireInWinter** [comment]
I think it is a separate issue that ErrorAction is ignored and that it still raises an exception, but I think the correct functionality of ErrorAction with SilentlyContinue should be to return no data, not to strip out some data. 

I think either Force or SkipIncompatibleProperties would be a better solution as you are explicitly telling it to get rid of data.

**Comment by mike-the-automator** [comment]
@FireInWinter  Created a separate issue as requested.  I think we are in agreement that the behavior requested in this issue should be implemented as a new parameter for the reasons already stated.  

I'm interested in contributing and would be happy to work on a PR for either issue, but I'd rather not invest time in this feature request if it has a low probability of being merged.  I'm not asking for a guarantee that my PR will be accepted, but it would be nice to know if I'm barking up the wrong tree here.

**Comment by bergmeister** [comment]
Can a maintainer please comment on the design. @steveleeMSFT ?

**Comment by SteveL-MSFT** [comment]
My preference would be to have a `-SkipUnsupportedTypes` switch.  `-Force` is too overloaded to me.  cc @joeyaiello @HemantMahawar for suggestions on the name of the switch.

**Comment by mklement0** [comment]
Generally, the use case you're describing is better handled using serialization based on CLIXML rather than JSON.

What makes this more cumbersome is the fact that you must export to a (temporary) _file_ first, using `Export-CliXml`, and also read from a file, using `Import-CliXml`; however, a new pair of `ConvertTo-CliXml` / `ConvertFrom-CliXml` cmdlets that directly output / read the XML as a strings are being worked on - see #3898

**Comment by SteveL-MSFT** [comment]
@mklement0 agree that if the use case is to serialize and deserialize later without care of the serialization format, CliXml is a better solution.  However, I can see cases where you want specifically json (more human readable than xml...)

**Comment by mklement0** [comment]
@SteveL-MSFT:

I merely restricted my comment to pointing out the CLIXML alternative, because I don't know what the right answer to the specific problem at hand is, but here are a few thoughts:

A simpler way to reproduce the problem, using a hashtable with a non-string key:

```powershell
@{ 1 = 'one' }  | ConvertTo-Json
```

The error message suggests that an explicit decision was made not to support non-string keys and to _fail_ in that event.

The alternatives - to be activated via new switch(es) - are:

* simply apply `.ToString()` to non-string keys and serialize nonetheless

* as suggested, _skip_ unserializable properties.

Given that JSON is not suitable for general-purpose serialization, however, I wonder if this is really necessary.
Generally, you will know what properties should be serialized ahead of time, and can filter out unsupported properties then, in the simplest case with `Select-Object -Property ...`

**Comment by iSazonov** [comment]
> I wonder if this is really necessary.

I wonder too. 
JSON is well standardized. I would have looked at what this standard says.

**Comment by markekraus** [comment]
> I wonder if this is really necessary.

Let's not let convenience become irrelevant in this project.

I have use cases for this myself that I have had to work around. I have object `A` that has nested property `B` that cannot be serialized, but I need to upload this to a generalized logging endpoint. that expects a JSON object. I don't care about  property 'B' even in the context of PowerShell or .NET. I have to create Object `C` that has everything but property `B` and then I can serialize it. Or I have to use a half dozen  other work arounds in this scenario including creatining my own JSON serializer.

**Comment by SteveL-MSFT** [comment]
This reminds me that in my [HttpListener module](https://www.powershellgallery.com/packages/HttpListener/1.0.2), I had to work around JSON serialization limitations by manually stripping types that were not supported.

**Comment by B-Art** [comment]
I would suggest a function which converts all none-string values to string values before piping it to convertto-json.
You can put a select statement in between for your final json output.

**Comment by mklement0** [comment]
@B-Art:

This issue is about _ignoring_ values that can't be converted to JSON.

By contrast, you're looking to for an opt-in method for _including_ such values by converting dictionary keys to strings, so I suggest you create a _new_ issue with your proposal.

As an aside, re your example: It isn't the `select` call that makes the command succeed, it is using `.GetEnumerator()`, because the latter sends a `[System.Collections.DictionaryEntry]` instance (rather than the whole hashtable) through the pipeline, which is itself not a dictionary, so the to-JSON conversion succeeds - however, that is _not_ the same as what `@{ '1' = 'one' } | ConvertTo-Json` would give you, i.e., the hashtable with stringified keys.

**Comment by heinrich-ulbricht** [comment]
Greeting from 5 years later where I need this functionality in a PowerShell Azure Function.

Specifically I'm trying to convert the output of "Get-Error" to JSON to push it to a logging endpoint. Fails with `ConvertTo-Json: The type 'System.Collections.ListDictionaryInternal' is not supported for serialization or deserialization of a dictionary. Keys must be strings.`.

**Comment by jelleholtkamp** [comment]
Would like to see this functionality as well. Need it in an Azure Pipeline for a similar scenario as @heinrich-ulbricht, to convert an error record to JSON for consumption further in the pipeline.

**Comment by B-Art** [comment]
`
1/0
$Error[0].InvocationInfo | ConvertTo-Json
`
gives:
`
{
  "MyCommand": null,
  "BoundParameters": {},
  "UnboundArguments": [],
  "ScriptLineNumber": 1,
  "OffsetInLine": 1,
  "HistoryId": -1,
  "ScriptName": "",
  "Line": "1/0",
  "PositionMessage": "At line:1 char:1\r\n+ 1/0\r\n+ ~~~",
  "PSScriptRoot": "",
  "PSCommandPath": null,
  "InvocationName": "",
  "PipelineLength": 0,
  "PipelinePosition": 0,
  "ExpectingInput": false,
  "CommandOrigin": 1,
  "DisplayScriptPosition": null
}
`

**Comment by DougChandler** [comment]
+1 for a parameter to `-SkipUnserializableTypes`. I am trying to serialize an exception to send to Log Analytics for further investigation and hitting the exception thrown by Convertto-json.

**Comment by iRon7** [comment]
@mike-the-automator,

> The property of the System.Exception that triggered this error was the "Data" field which I didn't really care about, but it prevented me from easily using the cmdlet to serialize the exception object.

I understand the need of a solution for the fact that a single item/property causes the whole `ConvertTo-Json` cmdlet to fail but I rather than a quick and dirty workaround, I would prefer it to (optionally) serialize as expected and convert keys that are not strings, like e.g. an integers (or anything else?)  to type cast to a string (`"$Key") considering that PowerShell is a loosely language and this does already works for:
```PowerShell
[PSCustomObject]@{ 1 = 'one' }  | ConvertTo-Json
{
  "1": "one"
}
```
And would of cause fail for something like:
```PowerShell
[PSCustomObject]@{ 1 = 'one'; '1' = 'one2' }  | ConvertTo-Json
ParserError:
Line |
   1 |  [PSCustomObject]@{ 1 = 'one'; '1' = 'one2' }  | ConvertTo-Json
     |                                ~~~
     | Duplicate keys '1' are not allowed in hash literals.
```
but that is the the actual error I would like to suppress...

**Comment by B-Art** [comment]
Something strange happens the other way around:
`'{ 1: "one", "1": "one2" }'  | Convertfrom-Json`
Will give:
```
1
-
one2
```
Or:
`'{ 1: "one", "1": "one2" }'  | Convertfrom-Json -AsHashtable`
Will give:
```
Name                           Value
----                           -----
1                              one2
```
So the first will be overwritten by the second. No errors...?

**Comment by sean-sauve** [comment]
Has anyone asked why ConvertTo-Json considers a ListDictionaryInternal as invalid?  It's entirely possible to convert the keys in a ListDictionaryInternal to JSON if GetEnumerator() is run on it first.

Transcript:

PS>1 / 0
RuntimeException: Attempted to divide by zero.
                                                                                                                                                                                                                                                           PS>$e = $error[0]
                                                                                                                                                                                                                                                                      PS>$e

RuntimeException: Attempted to divide by zero.
                                                                                                                                                                                                                                                                      PS>$e | ConvertTo-Json
WARNING: Resulting JSON is truncated as serialization has exceeded the set depth of 2.
>> TerminatingError(ConvertTo-Json): "The type 'System.Collections.ListDictionaryInternal' is not supported for serialization or deserialization of a dictionary. Keys must be strings."

ConvertTo-Json: The type 'System.Collections.ListDictionaryInternal' is not supported for serialization or deserialization of a dictionary. Keys must be strings.
                                                                                                                                                                                                                                                           PS>$e.Exception.Data | ConvertTo-Json
>> TerminatingError(ConvertTo-Json): "The type 'System.Collections.ListDictionaryInternal' is not supported for serialization or deserialization of a dictionary. Keys must be strings."

ConvertTo-Json: The type 'System.Collections.ListDictionaryInternal' is not supported for serialization or deserialization of a dictionary. Keys must be strings.
                                                                                                                                                                                                                                                           
PS>$e.Exception.Data.GetEnumerator() | % { $_ | ConvertTo-Json -Depth 2 } | Out-Null
WARNING: Resulting JSON is truncated as serialization has exceeded the set depth of 2.

**Comment by DougChandler** [comment]
Please only close it once it is resolved, otherwise it is still an issue.

**Comment by DougChandler** [comment]
Any progress on this issue?

**Comment by iSazonov** [comment]
We need to get WG conclusion.

Looking how #26521 add the `SkipUnsupportedTypes` switch I want ask - are there other types except IDictionary we need to take care of?
Looking how code of the cmdlet works I guess this type is the only one.

If this is an erroneous assumption, then why not add a parameter that will accept a list of types that should be ignored? (@SteveL-MSFT 's solution in his module (see above))
If this assumption is correct, then there are two options (@mklement0 's proposal):
 - By default (silently), convert keys to strings (add an index in case of conflicts). (And we can add a Verbose message too.)
- Or enable this with a more specific `DictionaryKeysToStrings` switch.

**Comment by yotsuda** [comment]
@iSazonov Thank you for your comments.

### Are there other types besides IDictionary?

As far as I confirmed, **IDictionary with non-string keys is the only type** that causes `ConvertTo-Json` to throw a terminating error. The `NonStringKeyInDictionary` error in [`ProcessDictionary`](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L761) is the only `ThrowTerminatingError` call in the serialization path.

So your assumption seems correct.

### Regarding the approach

Since IDictionary is the only problem type, I agree that a type list parameter would be over-engineering.

For @mklement0's two options:

1. **By default, convert keys to strings** (with Verbose message for conflicts)
2. **Enable with `-DictionaryKeysToStrings` switch**

I think option 2 (explicit switch) is safer for backward compatibility, as silently changing serialization behavior could surprise existing users.

My current PR uses a "skip" approach rather than "convert to string". If WG prefers the key conversion approach, I'm happy to update the PR accordingly.

Looking forward to WG's decision on which approach to take.

**Comment by iSazonov** [comment]
Looking our code I believe a root of the issue is that the code does a preprocessing of objects although both libraries (Newtonsoft.Json and System.Text.Json) are able to do IDictionary serialization.

Apparently, this preprocessing appeared to support PSObject serialization.

Also, Json is not a means of achieving readability, which is what YAML is for. 
Moreover, if we do not consider edge cases with keys that are converted to large strings, then this option is even preferable.

----

In fact, historically there have been several issues with this cmdlet. I don't think we should spend a lot of effort on fixing the old code.

I think we should go back to #11198, where we almost migrated to System.Text.Json. This PR was closed only because there were no resources to complete it.

If @yotsuda want to immerse yourself in this non-easy story, then you can grab this job. It was approved by PowerShell committee as an experimental feature, so I can promote it.

**Comment by DougChandler** [comment]
My preference is to fix it to serialise IDictionary objects rather than add a new parameter to work around it not working.
